README.md
# qrcode
A flutter plugin for scanning QR codes. Use AVCaptureSession in iOS and zxing in Android.

## Usage

### Use this package as a library

#### Add dependency

Add this to your package's pubspec.yaml file:

```dart
dependencies:
  qrcode:
    git:
      url: git://github.com/GrayLand119/qrcode.git
      path: qrcode
```

#### Install it

You can install packages from the command line:

with Flutter:

```
$ flutter pub get
```

#### Import it

Now in your Dart code, you can use:

```dart
import 'package:qrcode/qrcode.dart';
```

### Basic

```dart
class _MyAppState extends State<MyApp> {
  QRCaptureController _captureController = QRCaptureController();

  bool _isTorchOn = false;

  @override
  void initState() {
    super.initState();

    _captureController.onCapture((data) {
      print('onCapture----$data');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('扫一扫'),
      ),
      body: Stack(
        alignment: Alignment.center,
        children: <Widget>[
          Container(
            width: MediaQuery.of(context).size.width,
            height: MediaQuery.of(context).size.height,
            child: QRCaptureView(
              permissionAlertTitle: "提示",
              permissionAlertCancelTitle: "返回",
              permissionAlertContent: "需要访问相机来扫描二维码",
              permissionAlertOkTitle: "去设置",
              controller: _captureController,
            ),
          ),
          Padding(
            padding: EdgeInsets.symmetric(horizontal: 56),
            child: AspectRatio(
              aspectRatio: 264 / 258.0,
              child: Stack(
                alignment: _animation.value,
                children: <Widget>[
                  Image.asset('images/sao@3x.png'),
                  Image.asset('images/tiao@3x.png')
                ],
              ),
            ),
          ),
          Align(
            alignment: Alignment.bottomCenter,
            child: _buildToolBar(),
          ),
          Container(
            child: Text('$_captureText'),
          )
        ],
      ),
    );
  }
}
```

## Integration

### iOS
To use on iOS, you must add the following to your Info.plist


```
<key>NSCameraUsageDescription</key>
<string>Camera permission is required for qrcode scanning.</string>
<key>io.flutter.embedded_views_preview</key>
<true/>
```


qrcode/CHANGELOG.md
## 0.0.1

* init

## 1.0.2

* add torchMode setting

## 1.0.3

* fix frame setting on iOS

## 1.0.4

* add barCode scan for iOS

## 1.0.5

* iOS camer permission alert dialog can be customized.
* Permission alert dialog can jump to setting.

qrcode/lib/qrcode.dart
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';

typedef CaptureCallback(String data);

enum CaptureTorchMode { on, off }

class QRCaptureController {
  MethodChannel _methodChannel;
  CaptureCallback _capture;

  QRCaptureController();

  void _onPlatformViewCreated(int id) {
    _methodChannel = MethodChannel('plugins/qr_capture/method_$id');
    _methodChannel.setMethodCallHandler((MethodCall call) async {
      if (call.method == 'onCaptured') {
        if (_capture != null && call.arguments != null) {
          _capture(call.arguments.toString());
        }
      }
    });
  }

  void pause() {
    _methodChannel?.invokeMethod('pause');
  }

  void resume() {
    _methodChannel?.invokeMethod('resume');
  }

  void onCapture(CaptureCallback capture) {
    _capture = capture;
  }

  set torchMode(CaptureTorchMode mode) {
    var isOn = mode == CaptureTorchMode.on;
    _methodChannel?.invokeMethod('setTorchMode', isOn);
  }
}

class QRCaptureView extends StatefulWidget {
  final QRCaptureController controller;
  final String permissionAlertTitle;
  final String permissionAlertContent;
  final String permissionAlertCancelTitle;
  final String permissionAlertOkTitle;

  QRCaptureView(
      {Key key,
      this.controller,
      this.permissionAlertTitle,
      this.permissionAlertContent,
      this.permissionAlertCancelTitle,
      this.permissionAlertOkTitle})
      : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return QRCaptureViewState();
  }
}

class QRCaptureViewState extends State<QRCaptureView> {
  @override
  Widget build(BuildContext context) {
    if (Platform.isIOS) {
      return UiKitView(
        viewType: 'plugins/qr_capture_view',
        creationParams: {
          'permissionAlertTitle': widget.permissionAlertTitle,
          'permissionAlertContent': widget.permissionAlertContent,
          'permissionAlertCancelTitle': widget.permissionAlertCancelTitle,
          'permissionAlertOkTitle': widget.permissionAlertOkTitle,
        },
        creationParamsCodec: StandardMessageCodec(),
        onPlatformViewCreated: (id) {
          widget.controller._onPlatformViewCreated(id);
        },
      );
    } else {
      return AndroidView(
        viewType: 'plugins/qr_capture_view',
        creationParams: {
          'permissionAlertTitle': widget.permissionAlertTitle,
          'permissionAlertContent': widget.permissionAlertContent,
          'permissionAlertCancelTitle': widget.permissionAlertCancelTitle,
          'permissionAlertOkTitle': widget.permissionAlertOkTitle,
        },
        creationParamsCodec: StandardMessageCodec(),
        onPlatformViewCreated: (id) {
          widget.controller._onPlatformViewCreated(id);
        },
      );
    }
  }
}


qrcode/ios/Classes/QRCaptureView.m
//
//  QRCaptureView.m
//  Pods-Runner
//
//  Created by cdx on 2019/10/28.
//

#import "QRCaptureView.h"
#import <AVFoundation/AVFoundation.h>

@interface QRCaptureView () <AVCaptureMetadataOutputObjectsDelegate, FlutterPlugin>

@property(nonatomic, strong) AVCaptureSession *session;
@property(nonatomic, strong) FlutterMethodChannel *channel;
@property(nonatomic, weak) AVCaptureVideoPreviewLayer *captureLayer;
@property(nonatomic, strong) NSString *permissionAlertTitle;
@property(nonatomic, strong) NSString *permissionAlertContent;
@property(nonatomic, strong) NSString *permissionAlertCancelTitle;
@property(nonatomic, strong) NSString *permissionAlertOkTitle;

@end

@implementation QRCaptureView

- (AVCaptureSession *)session {
    if (!_session) {
        _session = [[AVCaptureSession alloc] init];
    }
    return _session;
}

- (instancetype)initWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if (self = [super initWithFrame:frame]) {
        NSString *name = [NSString stringWithFormat:@"plugins/qr_capture/method_%lld", viewId];
        FlutterMethodChannel *channel = [FlutterMethodChannel
                                         methodChannelWithName:name
                                         binaryMessenger:registrar.messenger];
        
        if (args != nil &&
            [args isKindOfClass:NSDictionary.class]) {
            NSDictionary *dict = (NSDictionary *)args;
            _permissionAlertTitle = dict[@"permissionAlertTitle"];
            _permissionAlertContent = dict[@"permissionAlertContent"];
            _permissionAlertCancelTitle = dict[@"permissionAlertCancelTitle"];
            _permissionAlertOkTitle = dict[@"permissionAlertOkTitle"];
        }
        
        if (!_permissionAlertTitle) {
            _permissionAlertTitle = @"Prompt";
        }
        if (!_permissionAlertCancelTitle) {
            _permissionAlertCancelTitle = @"Cancel";
        }
        if (!_permissionAlertOkTitle) {
            _permissionAlertOkTitle = @"Go Setting";
        }
        if (!_permissionAlertContent) {
            _permissionAlertContent = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"NSCameraUsageDescription"];
        }
        
        self.channel = channel;
        [registrar addMethodCallDelegate:self channel:channel];
        
        AVAuthorizationStatus status = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
        if(status == AVAuthorizationStatusAuthorized || status == AVAuthorizationStatusNotDetermined) {
            
            AVCaptureVideoPreviewLayer *layer = [AVCaptureVideoPreviewLayer layerWithSession:self.session];
            self.captureLayer = layer;
            
            layer.backgroundColor = [UIColor blackColor].CGColor;
            [self.layer addSublayer:layer];
            layer.videoGravity = AVLayerVideoGravityResizeAspectFill;
            
            AVCaptureDevice *device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
            AVCaptureDeviceInput *input = [[AVCaptureDeviceInput alloc] initWithDevice:device error:nil];
            AVCaptureMetadataOutput *output = [[AVCaptureMetadataOutput alloc] init];
            [self.session addInput:input];
            [self.session addOutput:output];
            self.session.sessionPreset = AVCaptureSessionPresetHigh;
            
            output.metadataObjectTypes = output.availableMetadataObjectTypes;
            [output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
            [output setMetadataObjectTypes:@[AVMetadataObjectTypeUPCECode, AVMetadataObjectTypeCode39Code, AVMetadataObjectTypeCode39Mod43Code,
            AVMetadataObjectTypeEAN13Code, AVMetadataObjectTypeEAN8Code, AVMetadataObjectTypeCode93Code, AVMetadataObjectTypeCode128Code,
            AVMetadataObjectTypePDF417Code, AVMetadataObjectTypeQRCode, AVMetadataObjectTypeAztecCode]];
            
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                  [self.session startRunning];
             });

        } else {
        
            UIAlertController *alertVC = [UIAlertController alertControllerWithTitle:_permissionAlertTitle message:_permissionAlertContent preferredStyle:UIAlertControllerStyleAlert];
            
            [alertVC addAction:[UIAlertAction actionWithTitle:_permissionAlertCancelTitle style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
                            
            }]];
            
            [alertVC addAction:[UIAlertAction actionWithTitle:_permissionAlertOkTitle style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                [[UIApplication sharedApplication]
                             openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
            }]];
            
            [UIApplication.sharedApplication.keyWindow.rootViewController presentViewController:alertVC animated:YES completion:nil];
        
            // Deprecated, iOS > 9.0
//            UIAlertView *alert = [[UIAlertView alloc] initWithTitle:_permissionAlertTitle message:_permissionAlertContent delegate:self cancelButtonTitle:_permissionAlertCancelTitle otherButtonTitles:_permissionAlertOkTitle, nil];
//                [alert show];
        }
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.captureLayer.frame = self.bounds;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:@"pause"]) {
        [self pause];
    } else if ([call.method isEqualToString:@"resume"]) {
        [self resume];
    } else if ([call.method isEqualToString:@"setTorchMode"]) {
        AVCaptureDevice *device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
        if (!device.hasTorch) {
            return;
        }
        NSNumber *isOn = call.arguments;
        [device lockForConfiguration:nil];
        if (isOn.boolValue) {
            [device setTorchMode:AVCaptureTorchModeOn];
        } else {
            [device setTorchMode:AVCaptureTorchModeOff];
        }
        [device unlockForConfiguration];
    }
}

+ (void)registerWithRegistrar:(nonnull NSObject<FlutterPluginRegistrar> *)registrar {}

- (void)resume {
    [self.session startRunning];
}

- (void)pause {
    [self.session stopRunning];
}

#pragma mark - AVCaptureMetadataOutputObjectsDelegate
-(void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection{
    if (metadataObjects.count>0) {
        AVMetadataMachineReadableCodeObject *metadataObject = metadataObjects[0];
        NSString *value = metadataObject.stringValue;
        if (value.length && self.channel) {
            [self.channel invokeMethod:@"onCaptured" arguments:value];
        }
    }
}

- (void)dealloc {
    [self.session stopRunning];
}

@end


qrcode/ios/Classes/QRCaptureViewFactory.m
//
//  QRCaptureViewFactory.m
//  Runner
//
//  Created by cdx on 2019/8/2.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

#import "QRCaptureViewFactory.h"
#import "QRCapturePlatformView.h"

@interface QRCaptureViewFactory ()
    
@property(nonatomic, strong) NSObject<FlutterPluginRegistrar>* registrar;
    
@end

@implementation QRCaptureViewFactory

- (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    self = [super init];
    if (self) {
        self.registrar = registrar;
    }
    return self;
}

    
- (nonnull NSObject<FlutterPlatformView> *)createWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args {
    return [[QRCapturePlatformView alloc] initWithFrame:frame viewIdentifier:viewId arguments:args registrar:self.registrar];
}

- (NSObject<FlutterMessageCodec> *)createArgsCodec {
//    return FlutterJSONMessageCodec.sharedInstance;
    return FlutterStandardMessageCodec.sharedInstance;
}
@end


qrcode/example/ios/Flutter/flutter_export_environment.sh
#!/bin/sh
# This is a generated file; do not edit or check into version control.
export "FLUTTER_ROOT=/Users/languilin/development/flutter"
export "FLUTTER_APPLICATION_PATH=/Users/languilin/WorkSpace/Private/GitHub/qrcode/qrcode/example"
export "FLUTTER_TARGET=/Users/languilin/WorkSpace/Private/GitHub/qrcode/qrcode/example/lib/main.dart"
export "FLUTTER_BUILD_DIR=build"
export "SYMROOT=${SOURCE_ROOT}/../build/ios"
export "OTHER_LDFLAGS=$(inherited) -framework Flutter"
export "FLUTTER_FRAMEWORK_DIR=/Users/languilin/development/flutter/bin/cache/artifacts/engine/ios"
export "FLUTTER_BUILD_NAME=1.0.0"
export "FLUTTER_BUILD_NUMBER=1"
export "DART_DEFINES=flutter.inspector.structuredErrors%3Dtrue"
export "DART_OBFUSCATION=false"
export "TRACK_WIDGET_CREATION=true"
export "TREE_SHAKE_ICONS=false"
export "PACKAGE_CONFIG=.packages"


qrcode/example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:qrcode/qrcode.dart';

void main() => runApp(GLMyApp());

class GLMyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyApp(),
    );
  }
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with TickerProviderStateMixin {
  QRCaptureController _captureController = QRCaptureController();
  Animation<Alignment> _animation;
  AnimationController _animationController;

  bool _isTorchOn = false;

  String _captureText = '';

  @override
  void initState() {
    super.initState();

    _captureController.onCapture((data) {
      print('onCapture----$data');
      setState(() {
        _captureText = data;
      });
    });

    _animationController = AnimationController(vsync: this, duration: Duration(seconds: 1));
    _animation = AlignmentTween(begin: Alignment.topCenter, end: Alignment.bottomCenter)
        .animate(_animationController)
          ..addListener(() {
            setState(() {});
          })
          ..addStatusListener((status) {
            if (status == AnimationStatus.completed) {
              _animationController.reverse();
            } else if (status == AnimationStatus.dismissed) {
              _animationController.forward();
            }
          });
    _animationController.forward();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('扫一扫'),
      ),
      body: Stack(
        alignment: Alignment.center,
        children: <Widget>[
          Container(
            width: MediaQuery.of(context).size.width,
            height: MediaQuery.of(context).size.height,
            child: QRCaptureView(
              permissionAlertTitle: "提示",
              permissionAlertCancelTitle: "返回",
              permissionAlertContent: "需要访问相机来扫描二维码",
              permissionAlertOkTitle: "去设置",
              controller: _captureController,
            ),
          ),
          Padding(
            padding: EdgeInsets.symmetric(horizontal: 56),
            child: AspectRatio(
              aspectRatio: 264 / 258.0,
              child: Stack(
                alignment: _animation.value,
                children: <Widget>[
                  Image.asset('images/sao@3x.png'),
                  Image.asset('images/tiao@3x.png')
                ],
              ),
            ),
          ),
          Align(
            alignment: Alignment.bottomCenter,
            child: _buildToolBar(),
          ),
          Container(
            child: Text('$_captureText'),
          )
        ],
      ),
    );
  }

  Widget _buildToolBar() {
    return Row(
      mainAxisSize: MainAxisSize.max,
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        FlatButton(
          onPressed: () {
            _captureController.pause();
          },
          child: Text('pause'),
        ),
        FlatButton(
          onPressed: () {
            if (_isTorchOn) {
              _captureController.torchMode = CaptureTorchMode.off;
            } else {
              _captureController.torchMode = CaptureTorchMode.on;
            }
            _isTorchOn = !_isTorchOn;
          },
          child: Text('torch'),
        ),
        FlatButton(
          onPressed: () {
            _captureController.resume();
          },
          child: Text('resume'),
        ),
      ],
    );
  }
}
